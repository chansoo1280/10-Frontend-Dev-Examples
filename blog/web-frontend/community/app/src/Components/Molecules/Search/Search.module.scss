@import "@Styles/scss/default.scss";

.search {
    display: inline-flex;
    position: relative;
    @include click-animating("");

    &__input {
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
        border-right: 0;
    }

    &__btn {
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
    }
}