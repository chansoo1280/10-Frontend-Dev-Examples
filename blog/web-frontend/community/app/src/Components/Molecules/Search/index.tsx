// #region Global Imports
import React, { ChangeEvent, ChangeEventHandler, MouseEventHandler, useRef, useState } from "react"
import classNames from "classnames"
// #endregion Global Imports

// #region Local Imports
import { defaultProps, Space, Input, Button, Icon } from "@Components"
import styles from "./Search.module.scss"
import useClickAnimating from "@Hooks/useClickAnimating"
// #endregion Local Imports

interface SearchProps extends defaultProps {
    value: string
    onChange: ChangeEventHandler<HTMLInputElement>
    onSearch: (value: string) => void
    placeholder?: string
    "data-click-animating"?: boolean
}

const Search = (props: SearchProps): JSX.Element => {
    const { value, show, className, onChange, onSearch, disabled, placeholder, ...rest } = props
    const [innerValue, setValue] = useState(value)
    const { isClick, execClickAnimation } = useClickAnimating()
    const prefixCls = "search"
    const classes = classNames(
        styles[`${prefixCls}`],
        {
            [styles[`${prefixCls}--hide`]]: show === false,
        },
        className,
    )
    const handleClick = () => {
        execClickAnimation()
    }
    const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
        onChange(e)
        setValue(e.currentTarget.value)
    }
    return (
        <Space onClick={handleClick} gap={0} padding={0} className={classes} {...rest} data-click-animating={isClick === true}>
            <Input placeholder={placeholder} disabled={disabled} value={innerValue} onChange={handleChange} className={classNames(styles[`${prefixCls}__input`])} />
            <Button disabled={disabled} onClick={() => onSearch(innerValue)} icon={<Icon iconName="xi-search" irText="검색" />} className={classNames(styles[`${prefixCls}__btn`])}></Button>
        </Space>
    )
}
export default Search
