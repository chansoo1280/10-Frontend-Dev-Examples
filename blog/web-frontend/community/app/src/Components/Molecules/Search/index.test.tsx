import { fireEvent, render, screen } from "@testing-library/react"
import Search from "./"
describe("Search", () => {
    it("renders search unchanged", () => {
        const handleChange = jest.fn()
        const handleSearch = jest.fn()
        const { container } = render(<Search value="" onChange={handleChange} onSearch={handleSearch} />)
        expect(container).toMatchSnapshot()
    })
    it("should render a search with the class", () => {
        const handleChange = jest.fn()
        const handleSearch = jest.fn()
        render(<Search placeholder="검색어를 입력해주세요" value="" onChange={handleChange} onSearch={handleSearch} data-testid="search" />)

        const search = screen.getByTestId("search")
        expect(search).toBeInTheDocument()
        expect(search).toHaveClass("search")
    })
    it("텍스트 입력 후 값 확인", () => {
        const handleChange = jest.fn()
        const handleSearch = jest.fn()
        render(<Search placeholder="검색어를 입력해주세요" value="" onChange={handleChange} onSearch={handleSearch} data-testid="search" />)

        const input = screen.getByPlaceholderText("검색어를 입력해주세요")
        fireEvent.change(input, { target: { value: "asdfasdf" } })
        expect(input).toHaveValue("asdfasdf")
        const button = screen.getByText("검색")
        fireEvent.click(button)
        expect(handleSearch).toBeCalledTimes(1)
    })
    it("비활성화 상태에서 텍스트 입력 후 값 확인", () => {
        const handleChange = jest.fn()
        const handleSearch = jest.fn()
        render(<Search disabled placeholder="검색어를 입력해주세요" value="" onChange={handleChange} onSearch={handleSearch} data-testid="search" />)

        const input = screen.getByPlaceholderText("검색어를 입력해주세요")
        fireEvent.change(input, { target: { value: "asdfasdf" } })
        expect(input).toHaveValue("")
    })
})
