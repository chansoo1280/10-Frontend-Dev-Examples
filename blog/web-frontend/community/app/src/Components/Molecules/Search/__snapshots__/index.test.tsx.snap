// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders search unchanged 1`] = `
<div>
  <div
    class="space search"
    data-click-animating="false"
    style="gap: 0; padding: 0px;"
  >
    <div
      class="input search__input"
      data-click-animating="false"
    >
      <input
        class="input__input"
        value=""
      />
    </div>
    <button
      class="btn btn--primary btn--icon-only search__btn"
      data-click-animating="false"
      type="button"
    >
      <i
        class="className xi-icon xi-search"
      >
        <span
          class="ir"
        >
          검색
        </span>
      </i>
    </button>
  </div>
</div>
`;
